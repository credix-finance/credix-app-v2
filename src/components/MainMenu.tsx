import { CredixLogo } from "@components/CredixLogo";
import axios from "axios";
import { User } from "pages";
import React, { useCallback, useEffect, useMemo, useState } from "react";
import create from "zustand";
import { SignInButton } from "./hackerhouse/SignInButton";
import { persist } from "zustand/middleware";

type Store = {
	user: User;
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	setUser: (u: User) => void;
	accessToken: string;
	setAccessToken: (t: string) => void;
};

export const useStore = create<Store>(
	persist(
		(set) => ({
			user: null,
			setUser: (user: User) => set((state) => ({ ...state, user: user })),
			accessToken: null,
			setAccessToken: (t: string) => set((state) => ({ ...state, accessToken: t })),
		}),
		{
			name: "storage",
		}
	)
);

export const MainMenu = () => {
	const [loggedIn, setLoggedIn] = useState<boolean>(false);
	const [started, setStarted] = useState<boolean>(false);
	const baseUrl = "http://127.0.0.1:8080";

	const { setUser, setAccessToken, accessToken } = useStore((state) => ({
		setUser: state.setUser,
		setAccessToken: state.setAccessToken,
		accessToken: state.accessToken,
	}));

	const client = useMemo(() => {
		console.log("creating client with token", accessToken);
		return axios.create({
			baseURL: baseUrl,
			headers: { Authorization: accessToken ? `Bearer ${accessToken}` : undefined },
		});
	}, [baseUrl, accessToken]);

	const getMe = useCallback(async () => {
		console.log("getting me");
		try {
			const response = await client.get("/api/users/me");
			setLoggedIn(true);
			console.log("user", response);
		} catch (e) {
			if (e.response && e.response.status === 401) {
				setLoggedIn(false);
				console.log("not logged in");
			}
		}
	}, [client]);

	const start = useCallback(async () => {
		await getMe();
		setStarted(true);
		console.log("started");
	}, [getMe]);

	useEffect(() => {
		if (!started) {
			console.log("starting");
			start();
		}
	}, [started, start]);

	const callback = (e, r) => {
		if (r) {
			client.defaults.headers["Authorization"] = `Bearer ${r.data.access_token}`;
			setLoggedIn(true);
			getMe();
			setAccessToken(r.data.access_token);
			console.log("logged in");
		}
	};

	return (
		<div className="w-full bg-neutral-0 flex justify-between items-center py-[16.5px] px-4">
			<div>
				<CredixLogo />
			</div>
			<div className="flex space-x-4">
				<SignInButton baseUrl="http://127.0.0.1:8080" isLoggedIn={loggedIn} onSignIn={callback} />
			</div>
		</div>
	);
};
