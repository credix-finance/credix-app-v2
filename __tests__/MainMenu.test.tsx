import React from "react";
import renderer from "react-test-renderer";
import { MainMenu } from "@components/MainMenu";
import { WalletProvider } from "@solana/wallet-adapter-react";
import { WalletModalProvider } from "@solana/wallet-adapter-react-ui";

test("Main menu", () => {
	const component = renderer.create(
		<WalletProvider wallets={[]} autoConnect>
			<WalletModalProvider>
				<MainMenu />
			</WalletModalProvider>
		</WalletProvider>
	);

	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
