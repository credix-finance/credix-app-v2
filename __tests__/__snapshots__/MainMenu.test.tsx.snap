// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main menu 1`] = `
<div
  className="w-full bg-credix-primary flex justify-between items-center py-[16.5px] px-4 lg:px-[88px]"
>
  <div>
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        className="w-7 h-10 text-black "
        data-name="Layer 1"
        fill="currentColor"
        id="Layer_1"
        viewBox="187.95 93.98 59.68 82.96"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M188,105.28c0-7.23,2.66-11.3,8.19-11.3,4.23,0,6.17,2.44,7.17,5.41l-3.16,1.07c-.66-2-1.56-3.57-4-3.57-3.09,0-4.44,2.53-4.44,6.54v3.69c0,4,1.35,6.54,4.44,6.54,2.48,0,3.38-1.56,4-3.56l3.16,1.06c-1,3-2.94,5.41-7.17,5.41C190.61,116.57,188,112.51,188,105.28Z"
        />
        <path
          d="M214.15,116.2h-3.48V94.36h8.17c4.13,0,6.41,2.5,6.41,6.57,0,3.31-1.68,5.63-4.72,6.13l5,9.14h-3.88l-4.64-8.83h-2.9Zm4.19-11.64c2.19,0,3.25-1,3.25-3v-1.34c0-2-1.06-3-3.25-3h-4.19v7.29Z"
        />
        <path
          d="M232.65,116.2V94.36h13.67V97.3H236.15v6.35H246v2.94h-9.8v6.67h10.17v2.94Z"
        />
        <path
          d="M188,129.34h6.51c5.53,0,8.23,3.82,8.23,10.92s-2.7,10.93-8.23,10.93H188Zm6.29,19c3.19,0,4.75-2.26,4.75-6.23v-3.63c0-4-1.56-6.23-4.75-6.23h-2.82v16.09Z"
        />
        <path
          d="M209.79,151.19v-2.66h5.14V132h-5.14v-2.66h13.77V132h-5.13v16.53h5.13v2.66Z"
        />
        <path
          d="M247.42,151.19h-4L241,146.87l-2.32-4.13h-.18l-2.38,4.13-2.57,4.32H229.8L236.49,140l-6.28-10.64h4l2.22,3.91,2.25,4h.16l2.25-4,2.32-3.91h3.72L240.78,140Z"
        />
        <rect
          height="3"
          width="59.68"
          x="187.95"
          y="173.94"
        />
      </svg>
    </a>
  </div>
  <div
    className="md:flex"
  >
    <button
      className="ant-btn ant-btn-primary ant-btn-lg rounded-[1px] border text-shadow-none shadow-none flex items-center justify-center gap-2 px-[25px] 
				bg-action-primary border-action-primary text-credix-primary
				hover:bg-neutral-60 hover:border-neutral-60
				active:bg-neutral-40 active:border-neutral-40 active:text-neutral-100
				disabled:bg-action-disable disabled:border-transparent disabled:text-disabled
			 text-lg font-semibold h-[50px] ml-4"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="w-6 h-6 fill-current"
        viewBox="0 0 24 25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M17 9.5H22V15.5H17C15.3431 15.5 14 14.1569 14 12.5C14 10.8431 15.3431 9.5 17 9.5ZM17 13.5C17.5523 13.5 18 13.0523 18 12.5C18 11.9477 17.5523 11.5 17 11.5C16.4477 11.5 16 11.9477 16 12.5C16 13.0523 16.4477 13.5 17 13.5Z"
          fillRule="evenodd"
        />
        <path
          clipRule="evenodd"
          d="M7.7587 4.5H16.2413C17.0463 4.49999 17.7106 4.49998 18.2518 4.54419C18.8139 4.59012 19.3306 4.68868 19.816 4.93597C20.5686 5.31947 21.1805 5.93139 21.564 6.68404C21.8113 7.16937 21.9099 7.68608 21.9558 8.24817C22 8.78936 22 9.45372 22 10.2587V14.7413C22 15.5463 22 16.2106 21.9558 16.7518C21.9099 17.3139 21.8113 17.8306 21.564 18.316C21.1805 19.0686 20.5686 19.6805 19.816 20.064C19.3306 20.3113 18.8139 20.4099 18.2518 20.4558C17.7106 20.5 17.0463 20.5 16.2413 20.5H7.75872C6.95374 20.5 6.28938 20.5 5.74817 20.4558C5.18608 20.4099 4.66937 20.3113 4.18404 20.064C3.43139 19.6805 2.81947 19.0686 2.43597 18.316C2.18868 17.8306 2.09012 17.3139 2.04419 16.7518C1.99998 16.2106 1.99999 15.5463 2 14.7413V10.2587C1.99999 9.45373 1.99998 8.78937 2.04419 8.24818C2.09012 7.68608 2.18868 7.16937 2.43597 6.68404C2.81947 5.93139 3.43139 5.31947 4.18404 4.93597C4.66937 4.68868 5.18608 4.59012 5.74817 4.54419C6.28937 4.49998 6.95373 4.49999 7.7587 4.5ZM5.91104 6.53755C5.47262 6.57337 5.24842 6.6383 5.09202 6.71799C4.7157 6.90973 4.40973 7.2157 4.21799 7.59202C4.1383 7.74842 4.07337 7.97262 4.03755 8.41104C4.00078 8.86113 4 9.44342 4 10.3V14.7C4 15.5566 4.00078 16.1389 4.03755 16.589C4.07337 17.0274 4.1383 17.2516 4.21799 17.408C4.40973 17.7843 4.7157 18.0903 5.09202 18.282C5.24842 18.3617 5.47262 18.4266 5.91104 18.4624C6.36113 18.4992 6.94342 18.5 7.8 18.5H16.2C17.0566 18.5 17.6389 18.4992 18.089 18.4624C18.5274 18.4266 18.7516 18.3617 18.908 18.282C19.2843 18.0903 19.5903 17.7843 19.782 17.408C19.8617 17.2516 19.9266 17.0274 19.9624 16.589C19.9992 16.1389 20 15.5566 20 14.7V10.3C20 9.44342 19.9992 8.86113 19.9624 8.41104C19.9266 7.97262 19.8617 7.74842 19.782 7.59202C19.5903 7.2157 19.2843 6.90973 18.908 6.71799C18.7516 6.6383 18.5274 6.57337 18.089 6.53755C17.6389 6.50078 17.0566 6.5 16.2 6.5H7.8C6.94342 6.5 6.36113 6.50078 5.91104 6.53755Z"
          fillRule="evenodd"
        />
      </svg>
      <span
        className="text-lg font-semibold"
      >
        Connect Wallet
      </span>
    </button>
  </div>
</div>
`;
